apps/aecore/src/aec_chain.erl

```diff
@@ -255,19 +256,53 @@ find_common_ancestor(Hash1, Hash2) when is_binary(Hash1), is_binary(Hash2) ->
 hash_is_connected_to_genesis(Hash) when is_binary(Hash) ->
     aec_chain_state:hash_is_connected_to_genesis(Hash).
 
--spec get_n_headers_from_hash(binary(), pos_integer()) -> {'ok', [aec_headers:header()]} |
-                                                          'error'.
-%% @doc Get n headers backwards in chain. Returns headers old -> new
-get_n_headers_from_hash(Hash, N) when is_binary(Hash), is_integer(N), N > 0 ->
-    get_n_headers_from_hash(get_header(Hash), N, []).
 
-get_n_headers_from_hash(_, 0, Acc) ->
+
+-spec get_at_most_n_headers_forward_from_hash(binary(), pos_integer()) ->
+                                                 {'ok', [aec_headers:header()]} |
+                                                 'error'.
+
+%%% @doc Get n headers forwards in chain. Returns headers old -> new
+%%%      This function is only defined for headers in the main chain to avoid
+%%%      ambiguity.
+get_at_most_n_headers_forward_from_hash(Hash, N) when is_binary(Hash), is_integer(N), N > 0 ->
+    case top_header() of
+        undefined -> error;
+        TopHeader ->
+            TopHeight = aec_headers:height(TopHeader),
+            case get_header(Hash) of
+                error -> error;
+                {ok, Header} ->
+                    Height = aec_headers:height(Header),
+                    Delta = min(TopHeight - Height, N - 1),
```

Comment:
I guess Delta could indeed be negative if there is a fork that is higher but has less difficulty

Reply:
Yes, exactly.